documentation/milestones/individual_milestones/developer_tools.md

```diff
@@ -0,0 +1,221 @@
+
+While enforcing compliance with these standards may increase the time and energy required for individual contributions, adopting an effective set of tools can increase the speed of delivery over time by reducing the overhead associated with reviewing new code and identifying potential bugs before they are deployed to production.
+
+### User stories
+
+- As a **full-time HHS staff member**, I want to:
+  - ensure that the codebase meets certain quality standards, so that it will be easier to onboard future maintainers and developers to the project
+  - have a mechanism for catching potential code issues during development or code review, so that we are not introducing bugs or security vulnerabilities in production
+- As an **open source contributor**, I want to:
+  - be able to reference documentation explaining how to use the developer tools, so that I don't have to learn how to use these tools on my own in order to contribute to the project
+  - have full test coverage for the codebase, so that I know when I've introduce code that changes or breaks existing behavior
+  - have code formatting and standards enforced with automated tooling, so that I don't have to learn and check that my code adheres to those standards manually
+  - be able to report security vulnerabilities to project maintainers directly, so that they can quickly create and deploy a fix before the vulnerability is made public
+- As a **maintainer of the project**, I want to:
+  - have code quality checks run automatically on each push, so that formatting, linting, or security issues are caught before being deployed to production
+  - see visible metrics and indications of code quality adherence such as badges on documentation and color coded check marks attached to code updates, so that I know all of the important checks are passing
+  - ensure that new contributions meet certain thresholds for test coverage, so that code contributions from internal and external developers also include tests which validate the code's behavior
+  - be automatically notified when updates are available for project dependencies, so that easily evaluate and adopt these new updates and ensure our dependencies don't become stale
+  - be notified when a security vulnerability is detected in our project or in an upstream dependency, so we can work to quickly address the vulnerability and deploy a fix
+  - securely manage and rotate keys and secrets related to the project, so that we minimize the risk of secrets being exposed or compromised
+
+## Technical description
+
+### Automated test framework
+
+Evaluate and adopt a framework to build and manage an automated test suite for the codebase. This test framework should include:
+
+- Unit tests
+- Integration tests
+- Test coverage
+
+### Code quality checks
+
+Evaluate and adopt a set of tools to enforce the following code quality checks:
+
+- Code Linting
```

Comment:
For future reference (not for this milestone but for the implementation / ADRs, so feel free to track this wherever), I like things like pre-commit hooks that do things like end-of-file fixing, sorting imports / removing unused imports, etc. Prettier for YAML / Markdown, immutability enforcement for at least JavaScript, enforcing typestrings for Python, docstring enforcement, tools like `darglint` that check whether docstring matches actual function, etc. 

I'm out of touch with what's trendy these days, so these are just ideas for consideration!

Reply:
I updated the Code Linting bullet with some of the specific suggestions you made. Many of the tools that we've selected (i.e. `ruff`, `black`) do incorporate some of these capabilities (i.e. unused import cleanup, sorting imports, EOF fixing, etc.) 

I hadn't heard of `darglint` but it sounds cool, unfortunately it looks like it's [no longer supported as a project](https://softwarerecs.stackexchange.com/questions/85405/is-there-a-successor-to-darglint). It looks like there might be some other projects that have sprung up since it was sunset though that we can investigate further.